File: frontend/src/app/model/ship.ts
export interface Model {
  id?: number;
  name?: string;
  color?: string;
}

export interface Owner {
  id?: number;
  name?: string;
}

export class Ship {
  id?: number;
  name?: string;

  // Velocidad (vector)
  vx?: number;
  vy?: number;

  // Posición actual en el tablero
  posX?: number;
  posY?: number;

  // Estado de la carrera
  turnCount?: number;
  racing?: boolean;
  finished?: boolean;

  // Relaciones con otros modelos (pueden ser objetos o strings/numbers)
  model?: Model | string | number;
  owner?: Owner | string | number;
}

export interface MoveDTO {
  deltaVx: number; // -1, 0, or 1
  deltaVy: number; // -1, 0, or 1
}

export interface PositionDTO {
  x: number;
  y: number;
  vx: number;
  vy: number;
  deltaVx: number;
  deltaVy: number;
  valid: boolean;
  invalidReason?: string;
}

export interface PossibleMovesDTO {
  currentX: number;
  currentY: number;
  currentVx: number;
  currentVy: number;
  possiblePositions: PositionDTO[];
}

File: backend/src/main/java/co/edu/javeriana/Proyecto_Web/model/Ship.java
package co.edu.javeriana.Proyecto_Web.model;

import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToOne;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Getter
@Setter
@NoArgsConstructor
public class Ship {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private long id;

    private String name;

    // Velocidad del barco (vector de velocidad)
    private int vx = 0;
    private int vy = 0;

    // Posición actual del barco en el tablero
    private int posX = 0;
    private int posY = 0;

    // Contador de turnos
    private int turnCount = 0;

    // Estado del barco en la carrera
    private boolean racing = false;
    private boolean finished = false;

    @ManyToOne
    private Model model;

    @OneToOne
    private Cell cell;

    @ManyToOne
    private User owner;

    public Ship(String name, Model model, User owner) {
        this.name = name;
        this.vx = 0;
        this.vy = 0;
        this.posX = 0;
        this.posY = 0;
        this.model = model;
        this.owner = owner;
    }

}
